### PR TITLE
refactor: fenced divs can be nested

### LINKED ISSUE #44: Fenced div blocks nesting
For recollection and possible future improvement, the `fenced_divs` extension support added in #36 is partial: it is doesn't support nesting of fenced blocks (as it always assumes `:::` as div marker -- For my excuse, I wasn't aware of that possibility, albeit very logical)

For parity with Pandoc, this should be supported, e.g.

```
::: { .diva }
I am in div A

:::: { .divb }
I am in nested div B
::::

Back to div A
:::
```

Generating, e.g. in HTML, something such as:

```
<div class="diva">
    <p>I am in div A</p>
    <div class="divb">
        <p>I am in nested div B</p>
    </div>
    <p>Back to div A</p>
</div>
```
(HTML indented for readability)

### PR DESCRIPTION
**Pull request by Omikhleia**
Closes #44

This repairs fenced divs nesting - I marked it as a "refactor" since 0.6.0 is not yet released, but I know there are various schools here, some would go for a "fix" anyhow...

This being said, this is more tricky than I thought... (Extended-)Markdown looks simple on the surface but is plenty of traps!
All our fenced blocks (i.e. tilde fenced code, backtick fenced code and colon fenced divs) currently use the same fencehead/fenceline/fencetail logic, and parse an "infostring" which will later be parsed as an attribute string. This seems pretty generic, but...

- Regarding the fenced div nesting issue...
  - For fenced code blocks, the closing fence must be at least as long as the opening fence (this is covered in the existing test)
  - For fenced divs however, my interpretation is that the closing fence must match the opening fence, since divs of increasing fence length can be nested.
  - **This is what I have fixed here**

- Regarding the infostring...
  - The existing tests for fenced code blocks have a test for "Info strings for backtick code blocks cannot contain backticks"
  - But as implemented (in the common infostring rule), it applies to all cases. Implying:
    - It applies to tilde fence code too. This might be wrong, see e.g. https://github.com/commonmark/commonmark-java/commit/62d5f1068fa3084a0f3757f54658e24336c6bd84, though it's a kind of an edge case... And an old unseen issue likely! To be clear, it means the following doesn't parse as a fenced code block (though it does in Pandoc):
      ```
      ~~~ info`
      Code
      ~~~
      ```
    - It applies to backtick/tilde code blocks and fenced divs with attributes, meaning, if I got that right, that one can't have a backtick in the `{ #id .class key="value" }` construct. This might perhaps be wrong, but it seems a rare edge case too (having a backtick in an identifier, a class or key value doesn't seem very useful anyway, does it?)
    - **I haven't tried to fix that here**: Whatever the standard or common interpretation say, I am not sure it's worth fixing (at the cost of loosing the current generic of fencehead/fenceline/fencetail and replacing them by very similar but slightly different methods for all cases. --> _Perhaps we should open a dedicated issue if it is however felt problematic?_

- Regarding paragraphing in fenced divs
  ```
  :::
  Line
  :::
  ```
  Generates `<div>Line</div>` while Pandoc gets `<div><p>Line</p></div>` -- **I haven't fixed that either here** --> _Perhaps worth an issue?_ (A work-around is to insert a blank line before the closing div fence)

### CODE REVIEW

**Comment by Witiko** [comment]
> It applies to backtick/tilde code blocks and fenced divs with attributes, meaning, if I got that right, that one can't have a backtick in the `{ #id .class key="value" }` construct

That seems wrong, because attributes are infostrings, even though we implemented them as such. In https://github.com/Witiko/markdown/pull/207, I added a separate pattern for fenced head with attributer rather than an infostring:

``` lua
parsers.fencehead_with_attributes
                     = function(char)
  return               C(parsers.nonindentspace) / function(s) fenceindent = #s end
                     * Cg(char^3, "fencelength")
                     * parsers.optionalspace * Ct(parsers.attributes)
                     * parsers.optionalspace * (parsers.newline + parsers.eof)
end
```

I added this pattern for use in divs, where there is no infostring, but it seems that we may want to use this pattern even for fenced code blocks to fix the backtick corner case.

> * The existing tests for fenced code blocks have a test for "Info strings for backtick code blocks cannot contain backticks"

Examples 145 and 146 of [the CommonMark spec][1] and the surrounding prose seem relevant here:

- [Info strings](https://spec.commonmark.org/0.30/#info-string) for backtick code blocks cannot contain backticks:
- [Info strings](https://spec.commonmark.org/0.30/#info-string) for tilde code blocks can contain backticks and tildes

However, note that lunamark does not comply with CommonMark in many places. I am planning to refactor witiko/markdown to support the full commonmark spec in the next major release, see https://github.com/Witiko/markdown/issues/212.

> Regarding paragraphing in fenced divs

This seems consistent with the behavior of all nested blocks such as blockquotes and backticks. Could this be a bug in Pandoc?

 [1]: https://spec.commonmark.org/0.30/#example-145

File: lunamark/reader/markdown.lua [hunk 1]
```diff
@@ -542,18 +544,23 @@ parsers.urlchar = parsers.anyescaped - parsers.newline - parsers.more
...
 
 parsers.BacktickFencedCodeBlock
-                     = parsers.fencehead(parsers.backtick)
+                     = parsers.fencehead(parsers.backtick, captures_geq_length)
                      * Cs(parsers.fencedline(parsers.backtick)^0)
                      * parsers.fencetail(parsers.backtick)
 
 parsers.ColonFencedDivBlock
-                     = parsers.fencehead(parsers.colon)
+                     = parsers.fencehead(parsers.colon, captures_equal_length)
```
Comment (Witiko) [inline lunamark/reader/markdown.lua:563]: Is this right? Pandoc is quite happy to parse the following:

``` md
::: foo
bar
:::::::
```

Comment (Witiko) [reply lunamark/reader/markdown.lua:563]: Here is the corresponding code in Pandoc's Haskell code:

https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Readers/Markdown.hs#L2066-L2089

I am not fluent in Haskell, but here is what I could glean:

- Pandoc does not care about the number of `:::` we use:

    ``` haskell
      string ":::"
      skipMany (char ':')
    ```

- Pandoc uses an entirely different logic for parsing fenced code and fenced divs. With fenced code, there is no internal structure, so counting backticks/tildes is the best we can hope for. With fenced divs, the inside of the div is Markdown text, so we can detect nesting without counting colons. In the Haskell code, this is tracked by the `stateFencedDivLevel` variable. This variable is then interspersed throughout the grammar.

**Review by Witiko — CHANGES_REQUESTED** [review]
Before this PR, divs could be nested similarly to fenced code by writing more (not fewer!) colons in the top-level div:

``` md
:::::::::::::: {.level1}
Fenced Divs can be nested

::: {.level2}
This is a nested Div
:::::::::::::
::::::::::::::::::::::::
```

As discussed in [my comment][1], Pandoc uses a more intricate logic for fenced divs, which essentially ignores the number of colons and allows you to nest divs even if you type fewer colons in the top-level div:

``` md
::: {.level1}
Fenced Divs can be nested

:::: {.level2}
This is a nested Div
::::
:::
```

It even allows you to do the following:



``` md
::: {.level1}
Fenced Divs can be nested

::: {.level2}
This is a nested Div
:::
:::
```

With this PR, we will correctly parse the second example. However, we will no longer correctly parse the first example and still not parse the third example. Therefore, this PR is not an improvement, although it has certainly unearthed many bugs and worms. 😉 

 [1]: https://github.com/jgm/lunamark/pull/46/files#r1020050463

**Comment by Witiko** [comment]
@jgm, can we please add @Omikhleia as a collaborator to jgm/lunamark?

**Comment by Witiko** [comment]
@Omikhleia Should we scrape the current implementation of nested divs and try again? Here is how I understand Pandoc's solution:

- Don't try to match the entire fenced div at once. Instead, recognize the beginning and the end of a fenced div separately.
- Refuse to match closing fences without corresponding opening fences. We can do this simply with a counter.

Conceptually, this seems even simpler than the current implementation and allows us to match nested fenced divs regardless of how many colons (`:`) we use. No need to consider the corner cases here: we track them in #48 and #49 and can tackle them later.

**Comment by Omikhleia** [comment]
@Witiko 
> Should we scrape the implementation of nested divs (...) and try again?
Well, if by "scrape" you mean reverting that part in the current code on master, maybe not necessarily -- It's wrong indeed, but at least in works in the simplest case (three symmetrical colons, no nesting). If you mean fixing it (say before 0.6.0 is tagged and released), yes, it surely has to... I'm not sure I can look at it soon, though - I'll have to be more careful, that time!

**Comment by Witiko** [comment]
I mean fixing it. There is no rush with the 0.6.0 release, but I will want the fenced divs ported to witiko/markdown before Christmas, so if you don't get to it by then, I will go ahead and fix it. 😉
